pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+    for (Map.Entry<String, Map<String, String>> entry : mapFields.entrySet()) {
+
+      String segmentName = entry.getKey();
+      Map<String, String> instanceStateMap = entry.getValue();
+
+      Map<String, String> newInstanceStateMap =
+          createNewInstanceStateMap(instanceStateMap, consumingServers, completedServersQueue);
+      if (MapUtils.isNotEmpty(newInstanceStateMap)) {
+        segmentNameToInstancesMap.put(segmentName, newInstanceStateMap);
+      }
+    }
+    return segmentNameToInstancesMap;
+  }
+
+  /**
+   * Given the instanceStateMap and a list of consuming and completed servers for a realtime resource,
+   * creates a new instanceStateMap, where one replica's instance is replaced from a consuming server to a completed server
+   * @param instanceStateMap
+   * @param consumingServers
+   * @param completedServersQueue
+   * @return
+   */
+  private Map<String, String> createNewInstanceStateMap(Map<String, String> instanceStateMap,
+      List<String> consumingServers, MinMaxPriorityQueue<Map.Entry<String, Integer>> completedServersQueue) {
+
+    Map<String, String> newInstanceStateMap = null;
+
+    // proceed only if all segments are ONLINE, and at least 1 server is from consuming list
+    for (String state : instanceStateMap.values()) {
+      if (!state.equals(PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE)) {
+        return newInstanceStateMap;
+      }
+    }
+
+    for (String instance : instanceStateMap.keySet()) {
+      if (consumingServers.contains(instance)) {
+        // Decide best strategy to pick completed server.
+        // 1. pick random from list of completed servers
+        // 2. pick completed server with minimum segments, based on ideal state of this resource
+        // 3. pick completed server with minimum segment, based on ideal state of all resources in this tenant
+        // 4. use SegmentAssignmentStrategy
+
+        // TODO: Using 2 for now. We should use 4. However the current interface and implementations cannot be used as is.
+        // We should refactor the SegmentAssignmentStrategy interface suitably and reuse it here
+
+        Map.Entry<String, Integer> completedInstance = completedServersQueue.pollFirst();
+
+        newInstanceStateMap = new HashMap<>(instanceStateMap.size());
+        newInstanceStateMap.putAll(instanceStateMap);
+        newInstanceStateMap.remove(instance);
+        newInstanceStateMap.put(completedInstance.getKey(),
+            PinotHelixSegmentOnlineOfflineStateModelGenerator.ONLINE_STATE);
+
+        completedInstance.setValue(completedInstance.getValue() + 1);
+        completedServersQueue.add(completedInstance);
+        break;
+      }
+    }
+    return newInstanceStateMap;
+  }
+
+  /**
+   * Updates the ideal state of given tablename with the new segment to instance state map
+   * @param tableNameWithType
+   * @param segmentNameToInstanceMap
+   */
+  private void updateIdealState(final String tableNameWithType,
```

Comment:
what happens when this fails? It probably fails because the version has changed. Retrying is not going to help without computing the new idealstate.

Reply:
+1, good catch

Reply:
Yes, thanks for catching that. Changed it to put the compute logic in retry along with the update